### PR TITLE
lookup configuration from file's path

### PR DESCRIPTION
**Pull request by gucong3000**
#1881 
continues to search configuration in these places all the way down the file tree until it finds acceptable configuration

This problem has been bothering me for a few months. I also do a lot of analysis work.
This problem will lead to all the tools that are developed based on stylelint, which must be resolved by itself to find the configuration file.
I think this is an obvious bug
Stylelint existing configuration file to find the logic is, from the process of the directory to start to find, if not, then to the directory tree search. And bug is that the starting point of the search should be the style file where the directory, not the process where the directory.

### CODE REVIEW

**Comment by evilebottnawi** [comment]
@gucong3000 need fix errors in builds

File: src/buildConfig.js [hunk 1]
```diff
@@ -44,7 +46,7 @@ export default function (options) {
   const configBasedir = options.configBasedir || false
 
   if (rawConfig) {
-    const configDir = configBasedir || process.cwd()
+    const configDir = configBasedir || (options.filepath ? path.dirname(options.filepath) : process.cwd())
```
Comment (Arcanemagus) [inline src/buildConfig.js]: Where is `options.filepath` coming from? Why isn't it in camelCase?

[The Documentation](http://stylelint.io/user-guide/node-api/#configfile) already says the path is based on the `cwd()`, but this PR doesn't update it to account for this changed behavior.

This completely ignores the `codeFilename` value from the Node API.

File: src/buildConfig.js [hunk 2]
```diff
@@ -119,7 +123,7 @@ function addIgnores(config, configDir, ignorePath) {
   // Absoluteize config.ignoreFiles
   if (config.ignoreFiles) {
     config.ignoreFiles = [].concat(config.ignoreFiles).map(glob => {
-      if (path.isAbsolute(glob)) return glob
+      if (path.isAbsolute(glob.replace(/^\!/, ""))) return glob
```
Comment (Arcanemagus) [inline src/buildConfig.js]: What is the point of this change? It doesn't seem related to anything else in this PR...

**Comment by Arcanemagus** [comment]
What does this PR do that `configBasedir` doesn't already accomplish?

File: src/buildConfig.js (hunk 1 above)
Comment (gucong3000) [inline src/buildConfig.js]: In Atom `linter-stylelint` and some third party utilities that run `stylelint` will pass a `cosmiconfig` result object to `stylelint`. result.filepath is configuration file's path

If a `options` is `cosmiconfig` result object, we shold do logic same with this part of code:

```
  return cosmiconfig("stylelint", cosmiconfigOptions).then(result => {
    if (!result) throw configurationError("No configuration found")
    rootConfigDir = configBasedir || path.dirname(result.filepath)
    return augmentConfig(result.config, rootConfigDir, {
      addIgnorePatterns: true,
      ignorePath: options.ignorePath,
    })
  })
```
`rootConfigDir = configBasedir || path.dirname(result.filepath)`

File: src/buildConfig.js (hunk 2 above)
Comment (gucong3000) [inline src/buildConfig.js]: `path.isAbsolute` will return `false`, when `glob === "!E:\myproj\**\*.css"`

In some case of `/stylelint/src/__tests__/ignore-test.js`, one configuration object will use for two files, so this object will pass to `buildConfig` twice.
In first time `config.ignoreFiles` will convert to absolute paths.
In second time `config.ignoreFiles` maybe have "!E:\myproj\**\*.css"`.

**Comment by gucong3000** [comment]
@evilebottnawi
> need fix errors in builds

there is no any error in build job, It is just travis-ci happend  a network error

File: src/buildConfig.js (hunk 1 above)
Comment (Arcanemagus) [inline src/buildConfig.js]: There is no file name with the Node.js API, the block of code is being passed in directly. That block of code can be associated with a file name using the `codeFilename` option.

**Comment by davidtheclark** [comment]
@gucong3000 Just a short message to let you know that I'm back from my trip and this is on my mind. I was going to tackle your other issue first (expose a method for looking up configuration), and it will affect the implementation of this.

**Comment by davidtheclark** [comment]
@gucong3000 Thanks for all the work here! I'm taking it as inspiration for a larger refactor in https://github.com/stylelint/stylelint/pull/1930, which will also close #1882 when it's done.

**Comment by davidtheclark** [comment]
This work is incorporated into #1930. Thanks again @gucong3000.
